Enhancement/10059 refactor not set

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #10059 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on composer.json:
```diff
@@ -46,11 +47,17 @@
+    "wordpress-install-dir": "vendor/roots/wordpress",
+    "patches": {
+      "google/apiclient-services": {
+        "Fixes for empty filter": "patches/google-api-client-services.diff"
```
Hmm, we already have a file in `patches` called `@wordpress+data+4.27.3.patch`, indicating it's a patch for `@wordpress/data` version `4.27.3`.

Can we rename the file we're introducing for consistency, so this it includes the version of `google-api-client-services` that it's patching (`v0.355.0`), and has the suffix `patch` instead of `diff`?

```suggestion
        "Fixes for empty filter": "patches/google-api-client-services-v0.355.0.patch"
```

techanvil reviewed (changes requested):
Thanks @ankitrox, this is looking good. It needs a few updates, please see my comments.

Please can you also add a QAB?

ankitrox commented:
Thank you for reviewing the PR and adding your valuable insights @techanvil. I have updated `notExpression` to `true` and also updated the code as per your suggestions.

QAB has been added into the issue.

Requesting another review for the changes made.

Note: It seems like JS tests are failing in pipeline, but those are all passing locally, so moving this to CR.

techanvil commented on assets/js/modules/analytics-4/hooks/useAudienceTilesReports.js:
```diff
@@ -296,6 +296,12 @@ export default function useAudienceTilesReports( {
 		startDate,
 		endDate,
 		dimensions: [ 'city' ],
+		dimensionFilters: {
+			city: {
+				filterType: 'emptyFilter',
+				notExpression: true,
+			},
+		},
```
We don't actually want to use the `emptyFilter` filter here, because we are deliberately doing our own JS-side filtering for the "top cities" metric to make sure the percentages are calculated correctly and match the "Top Cities" KM widget.

```suggestion
```

Note how we use the `reportRowsWithSetValues()` for the metric in both the Audience Tile and the KM tile; these will be addressed in https://github.com/google/site-kit-wp/issues/10143:

https://github.com/google/site-kit-wp/blob/2067156be2d35584b7b77205a1cf5e70b984c342/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/AudienceTiles.js#L571-L574

https://github.com/google/site-kit-wp/blob/2067156be2d35584b7b77205a1cf5e70b984c342/assets/js/modules/analytics-4/components/widgets/TopCitiesWidget.js#L120

techanvil commented on assets/js/modules/analytics-4/hooks/useAudienceTilesReports.js:
```diff
@@ -305,7 +311,7 @@ export default function useAudienceTilesReports( {
 				desc: true,
 			},
 		],
-		limit: 4, // Limit is set to 4 so that (not set) can be filtered out and 3 cities remain to display.
+		limit: 3,
```
As above, please revert this change.

```suggestion
		limit: 4, // Limit is set to 4 so that (not set) can be filtered out and 3 cities remain to display.
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/index.stories.js:
```diff
@@ -66,6 +66,12 @@ const topCitiesReportOptions = {
 	endDate: '2024-03-27',
 	startDate: '2024-02-29',
 	dimensions: [ 'city' ],
+	dimensionFilters: {
+		city: {
+			filterType: 'emptyFilter',
+			notExpression: true,
+		},
+	},
```
This will need reverting as per my comment on `assets/js/modules/analytics-4/hooks/useAudienceTilesReports.js`.

```suggestion
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/index.stories.js:
```diff
@@ -75,7 +81,7 @@ const topCitiesReportOptions = {
 		},
 	],
-	limit: 4,
+	limit: 3,
```
As above:

```suggestion
	limit: 4,
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/index.stories.js:
```diff
@@ -602,6 +611,10 @@ SingleTileErrored.args = {
 			endDate,
 			dimensions: [ 'city' ],
 			dimensionFilters: {
+				city: {
+					filterType: 'emptyFilter',
+					notExpression: true,
+				},
```
As above:

```suggestion
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/index.stories.js:
```diff
@@ -613,7 +626,7 @@ SingleTileErrored.args = {
 				},
 			],
-			limit: 4,
+			limit: 3,
```
As above:

```suggestion
			limit: 4,
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/index.test.js:
```diff
@@ -158,6 +158,12 @@ function provideAudienceTilesMockReport(
 		startDate,
 		endDate,
 		dimensions: [ 'city' ],
+		dimensionFilters: {
+			city: {
+				filterType: 'emptyFilter',
+				notExpression: true,
+			},
+		},
```
As above:

```suggestion
```

techanvil commented on assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget/index.test.js:
```diff
@@ -167,7 +173,7 @@ function provideAudienceTilesMockReport(
 			},
 		],
-		limit: 4,
+		limit: 3,
```
As above:

```suggestion
		limit: 4,
```

techanvil reviewed (changes requested):
Thanks @ankitrox, this is nearly there, but the changes to the "top cities" report args needs to be reverted.

ankitrox commented:
@techanvil Thanks for looking into the changes. I have reverted the changes in City widget for AS.

Requesting another review to look into overall status of the changeset.

Thank you.

techanvil reviewed (approved):
Thanks @ankitrox! This is LGTM.

Please note, I made a couple of small tweaks:

- Removed an unnecessary change to the `AudienceTilesWidget` stories:  https://github.com/google/site-kit-wp/pull/10196/commits/6cd97e0b7ce9797dafe765607dcee3474e5b4af3
- Updated the QAB to refer to the correct list of "top cities" tiles/metrics.

Note that I've confirmed the failing E2E test is unrelated to the changes in this PR.
